util/setuidgid.c

```diff
@@ -0,0 +1,332 @@
+
+/**
+ * setgroups_none():
+ * Attempt to leave all supplementary groups.  If we do not know how to do this
+ * on the platform, return 0 anyway.
+ */
+static int
+setgroups_none()
+{
+
+#ifndef NO_SETGROUPS
+	if (setgroups(0, NULL)) {
+		/* We handle EPERM separately; keep it in errno. */
+		if (errno != EPERM)
+			warnp("setgroups()");
+		goto err0;
+	}
+#endif
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Check if we're in any supplementary groups. */
+static int
+check_supplementary_groups_none()
+{
+	gid_t grouplist[1];
+	int ngroups;
+
+	/* Find number of groups. */
+	if ((ngroups = getgroups(0, NULL)) < 0) {
+		warnp("getgroups()");
+		goto err0;
+	}
+
+	/* Check group membership. */
+	if (ngroups > 1)
+		/* Definitely failed. */
+		goto err0;
+	if (ngroups == 1) {
+		/*
+		 * POSIX allows getgroups() to return the effective group ID,
+		 * so if we're in exactly one group, we need to check that GID.
+		 */
+		if (getgroups(1, grouplist) != 1) {
+			warnp("getgroups()");
+			goto err0;
+		}
+		if (grouplist[0] != getegid())
+			goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Set the effective GID to the group represented by ${groupname}. */
+static int
+set_group(const char * groupname)
+{
+	gid_t gid;
+	struct group * group_info;
+
+	/* Find numerical GID. */
+	errno = 0;
+	if ((group_info = getgrnam(groupname)) == NULL) {
+		if (errno)
+			warnp("getgrnam(\"%s\")", groupname);
+		else
+			warn0("No such group: %s", groupname);
+		goto err0;
+	}
+	gid = group_info->gr_gid;
+
+	/* Set GID. */
+	if (setgid(gid)) {
+		/* We handle EPERM separately; keep it in errno. */
+		if (errno != EPERM)
+			warnp("setgid(%lu)", (unsigned long int)gid);
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Set the effective UID to the user represented by ${username}. */
+static int
+set_user(const char * username)
+{
+	uid_t uid;
+	struct passwd * user_info;
+
+	/* Calculate numerical UID. */
+	errno = 0;
+	if ((user_info = getpwnam(username)) == NULL) {
+		if (errno)
+			warnp("getpwnam(\"%s\")", username);
+		else
+			warn0("No such user: %s", username);
+		goto err0;
+	}
+	uid = user_info->pw_uid;
+
+	/* Set UID. */
+	if (setuid(uid)) {
+		/* We handle EPERM separately; keep it in errno. */
+		if (errno != EPERM)
+			warnp("setuid(%lu)", (unsigned long int)uid);
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Parses "username", ":groupname", or "username:groupname". */
+static int
+string_extract_user_group(const char * combined, char ** username_p,
+    char ** groupname_p)
+{
+	size_t pos;
+	size_t len;
+
+	/* Sanity check. */
+	assert(combined != NULL);
+
+	/* Search for ":" and get string length. */
+	pos = strcspn(combined, ":");
+	len = strlen(combined);
+
+	/* Reject silly strings. */
+	if (pos == len - 1) {
+		warn0("Groupname is blank.");
```

Comment:
I'd go with
```
warn0("Empty group name: %s", combined);
```
here (both the spelling of "group name" and printing the failed string).